### ISSUE TITLE
Typescript definitions

### ISSUE THREAD

**Issue by thellimist** [raycharius/slack-block-builder#24]
Would be great to have "TypeScript type definitions.". 

Thanks

**Comment by raycharius** [comment]
Thanks! Not sure I have the time in the next month to do these, but if you feel up to it and need it before, feel free to send over a PR!

**Comment by karoun** [comment]
```typescript
declare module 'slack-block-builder' {
  // https://github.com/sindresorhus/type-fest/blob/e6827d94e3af8b5611cf330263df53f62ce683d6/source/opaque.d.ts
  type Opaque<Type, Token = unknown> = Type & { readonly __opaque__: Token };

  declare namespace SlackBlockBuilder {
    declare namespace Bit {
      interface Bit {
        end(): this;
      }

      interface AttachmentParams {
        color?: string;
      }

      interface Attachment extends Bit {
        new (params?: AttachmentParams): Attachment;
        color(color: string): this;
        blocks(...x0: Array<Block.Block | Block.Block[]>): this;
      }

      interface OptionGroupParams {
        label?: string;
      }

      interface OptionGroup extends Bit {
        new (params?: OptionGroupParams): OptionGroup;
        label(label: string): this;
        options(...x0: Array<Option | Option[]>): this;
      }

      interface OptionParams {
        text?: string;
        value?: string;
        description?: string;
        url?: string;
      }

      interface Option extends Bit {
        new (params?: OptionParams): Option;
        text(text: string): this;
        value(value: string): this;
        description(description: string): this;
        url(url: string): this;
      }

      interface ConfirmationDialogParams {
        title?: string;
        text?: string;
        confirm?: string;
        deny?: string;
      }

      interface ConfirmationDialog extends Bit {
        new (params?: ConfirmationDialogParams): ConfirmationDialog;
      }

      export function Attachment(params?: AttachmentParams): Attachment;
      export function ConfirmationDialog(
        params?: ConfirmationDialogParams
      ): ConfirmationDialog;
      export function Option(params?: OptionParams): Option;
      export function OptionGroup(params?: OptionGroupParams): OptionGroup;
    }

    declare namespace Block {
      interface Block {
        blockId(blockId: string): this;
        end(): this;
      }

      interface ActionsParams {
        blockId?: string;
      }

      interface Actions extends Block {
        new (params?: ActionsParams): Actions;
        elements(...x0: Array<Element.Element | Element.Element[]>): this;
      }

      interface ContextParams {
        blockId?: string;
      }

      interface Context extends Block {
        new (params?: ContextParams): Context;
        elements(...x0: Array<Element.Element | Element.Element[]>): this;
      }

      interface DividerParams {
        blockId?: string;
      }

      interface Divider extends Block {
        new (params?: DividerParams): Divider;
      }

      interface FileParams {
        blockId?: string;
        externalId?: string;
      }

      interface File extends Block {
        new (params?: FileParams): File;
        externalId(externalId: string): this;
      }

      interface HeaderParams {
        blockId?: string;
        text?: string;
      }

      interface Header extends Block {
        new (params?: HeaderParams): Header;
        text(text: string): this;
      }

      interface ImageParams {
        blockId?: string;
        imageUrl?: string;
        altText?: string;
        title?: string;
      }

      interface Image extends Block {
        new (params?: ImageParams): Image;
        imageUrl(imageUrl: string): this;
        altText(altText: string): this;
        title(title: string): this;
      }

      interface InputParams {
        blockId?: string;
        label?: string;
        hint?: string;
      }

      interface Input extends Block {
        new (params?: InputParams): Input;
        label(label: string): this;
        element(element: Element.Element): this;
        hint(hint: string): this;
        optional(): this;
      }

      interface SectionParams {
        blockId?: string;
        text?: string;
      }

      interface Section extends Block {
        new (params?: SectionParams): Section;
        text(text: string): this;
        fields(...fields: Array<string | string[]>): this;
        accessory(element: Element.Element): this;
      }
    }

    declare namespace Element {
      interface Element {
        end(): this;
      }

      interface ActionElement extends Element {
        actionId(actionId: string): this;
      }

      interface ConfirmableElement extends ActionElement {
        confirm(obj: Bit.ConfirmationDialog): this;
      }

      interface SelectableElement extends ConfirmableElement {
        options(...x0: Array<Bit.Option | Bit.Option[]>): this;
      }

      interface SelectElement extends ConfirmableElement {
        placeholder(placeholder: string): this;
      }

      interface MultiSelectElement extends SelectElement {
        maxSelectedItems(maxSelectedItems: number): this;
      }

      interface ButtonParams {
        actionId?: string;
        text?: string;
        url?: string;
        value?: string;
      }

      interface Button extends ConfirmableElement {
        new (params?: ButtonParams): Button;
        text(text: string): this;
        url(url: string): this;
        value(value: string): this;
        primary(): this;
        danger(): this;
      }

      interface ChannelMultiSelectParams {
        actionId?: string;
        placeholder?: string;
        maxSelectedItems?: number;
      }

      interface ChannelMultiSelect extends MultiSelectElement {
        new (params?: ChannelMultiSelectParams): ChannelMultiSelect;
        initialChannels(...strings: Array<string | string[]>): this;
      }

      interface ChannelSelectParams {
        actionId?: string;
        placeholder?: string;
        initialChannel?: string;
      }

      interface ChannelSelect extends SelectElement {
        new (params?: ChannelSelectParams): ChannelSelect;
        initialChannel(initialChannel: string): this;
        responseUrlEnabled(): this;
      }

      interface CheckboxesParams {
        actionId?: string;
      }

      interface Checkboxes extends SelectableElement {
        new (params?: CheckboxesParams): Checkboxes;
        options(...x0: Array<Bit.Option | Bit.Option[]>): this;
      }

      interface ConversationMultiSelectParams {
        actionId?: string;
        placeholder?: string;
        maxSelectedItems?: number;
      }

      interface ConversationMultiSelect extends MultiSelectElement {
        new (params?: ConversationMultiSelectParams): ConversationMultiSelect;
        initialConversations(...strings: Array<string | string[]>): this;
        defaultToCurrentConversation(): this;
        filter(...filters: Array<string | string[]>): this;
        excludeExternalSharedChannels(): this;
        excludeBotUsers(): this;
      }

      interface ConversationSelectParams {
        actionId?: string;
        placeholder?: string;
        initialConversation?: string;
      }

      interface ConversationSelect extends SelectElement {
        new (params?: ConversationSelectParams): ConversationSelect;
        initialConversation(initialConversation: string): this;
        defaultToCurrentConversation(): this;
        responseUrlEnabled(): this;
        filter(...filters: Array<string | string[]>): this;
        excludeExternalSharedChannels(): this;
        excludeBotUsers(): this;
      }

      interface DatePickerParams {
        actionId?: string;
        placeholder?: string;
        initialDate?: Date;
      }

      interface DatePicker extends ConfirmableElement {
        new (params?: DatePickerParams): DatePicker;
        placeholder(placeholder: string): this;
        initialDate(date: Date): this;
      }

      interface ExternalMultiSelectParams {
        actionId?: string;
        placeholder?: string;
        maxSelectedItems?: number;
        minQueryLength?: number;
      }

      interface ExternalMultiSelect extends MultiSelectElement {
        new (params?: ExternalMultiSelectParams): ExternalMultiSelect;
        minQueryLength(minQueryLength: number): this;
        initialOptions(...options: Array<Bit.Option | Bit.Option[]>): this;
      }

      interface ExternalSelectParams {
        actionId?: string;
        placeholder?: string;
        minQueryLength?: number;
      }

      interface ExternalSelect extends SelectElement {
        new (params?: ExternalSelectParams): ExternalSelect;
        initialOption(option: Bit.Option): this;
        minQueryLength(minQueryLength: number): this;
      }

      interface ImgParams {
        imageUrl?: string;
        altText?: string;
      }

      interface Img extends Element {
        new (params?: ImgParams): Img;
        imageUrl(imageUrl: string): this;
        altText(altText: string): this;
      }

      interface OverflowMenuParams {
        actionId?: string;
      }

      interface OverflowMenu extends SelectableElement {
        new (params?: OverflowMenuParams): OverflowMenu;
      }

      interface RadioButtonsParams {
        actionId?: string;
      }

      interface RadioButtons extends SelectableElement {
        new (params?: RadioButtonsParams): RadioButtons;
        initialOption(option: Bit.Option): this;
      }

      interface StaticMultiSelectParams {
        actionId?: string;
        placeholder?: string;
        maxSelectedItems?: number;
      }

      interface StaticMultiSelect extends MultiSelectElement {
        new (params?: StaticMultiSelectParams): StaticMultiSelect;
        options(...x0: Array<Bit.Option | Bit.Option[]>): this;
        optionGroups(...x0: Array<Bit.OptionGroup | Bit.OptionGroup[]>): this;
        initialOptions(...x0: Array<Bit.Option | Bit.Option[]>): this;
      }

      interface StaticSelectParams {
        actionId?: string;
        placeholder?: string;
      }

      interface StaticSelect extends SelectElement {
        new (params?: StaticSelectParams): StaticSelect;
        options(...x0: Array<Bit.Option | Bit.Option[]>): this;
        optionGroups(...x0: Array<Bit.OptionGroup | Bit.OptionGroup[]>): this;
        initialOption(option: Bit.Option): this;
      }

      interface TextInputParams {
        actionId?: string;
        placeholder?: string;
        initialValue?: string;
        minLength?: number;
        maxLength?: number;
      }

      interface TextInput extends ActionElement {
        new (params?: TextInputParams): TextInput;
        placeholder(placeholder: string): this;
        initialValue(initialValue: string): this;
        multiline(): this;
        minLength(minLength: number): this;
        maxLength(maxLength: number): this;
      }

      interface UserMultiSelectParams {
        actionId?: string;
        placeholder?: string;
        maxSelectedItems?: number;
      }

      interface UserMultiSelect extends MultiSelectElement {
        new (params?: UserMultiSelectParams): UserMultiSelect;
        maxSelectedItems(maxSelectedItems: number): this;
      }

      interface UserSelectParams {
        actionId?: string;
        placeholder?: string;
        initialUser?: string;
      }

      interface UserSelect extends SelectElement {
        new (params?: UserSelectParams): UserSelect;
        initialUser(initialUser: string): this;
      }
    }

    declare namespace Surface {
      // We want to ensure the only way to get a value of type
      // BuiltObject is by calling buildToObject(), so we use
      // an opaque type which cannot be constructed by a user.
      type BuiltObject = Opaque<unknown, 'BuiltObject'>;

      interface Surface {
        blocks(...x0: Array<Block.Block | Block.Block[]>): this;
        buildToObject(): BuiltObject;
        buildToJSON(): string;
        getBlocks(): Block.Block[];
      }

      interface AdvancedSurface extends Surface {
        privateMetaData(privateMetaData: string): this;
        callbackId(callbackId: string): this;
        externalId(externalId: string): this;
        printPreviewUrl(): void;
      }

      interface HomeTabParams {
        callbackId?: string;
        externalId?: string;
        privateMetaData?: string;
      }

      interface HomeTab extends AdvancedSurface {
        new (params?: HomeTabParams): HomeTab;
      }

      interface MessageParams {
        channel?: string;
        text?: string;
        threadTs?: string;
        postAt?: string;
      }

      interface Message extends Surface {
        new (params?: MessageParams): Message;
        attachments(...x0: Array<Bit.Attachment | Bit.Attachment[]>): this;
        channel(channel: string): this;
        text(text: string): this;
        asUser(): this;
        threadTs(threadTs: string): this;
        ts(ts: string): this;
        replaceOriginal(): this;
        deleteOriginal(): this;
        ephemeral(): this;
        inChannel(): this;
        postAt(timestamp: string): this;
        getAttachments(): Bit.Attachment[];
      }

      interface ModalParams {
        title?: string;
        close?: string;
        submit?: string;
        callbackId?: string;
        externalId?: string;
        privateMetaData?: string;
      }

      interface Modal extends AdvancedSurface {
        new (params?: ModalParams): Modal;
        title(title: string): this;
        close(close: string): this;
        submit(submit: string): this;
        clearOnClose(): this;
        notifyOnClose(): this;
      }
    }

    export const Bits: {
      Attachment(params?: Bit.AttachmentParams): Bit.Attachment;
      ConfirmationDialog(
        params?: Bit.ConfirmationDialogParams
      ): Bit.ConfirmationDialog;
      Option(params?: Bit.OptionParams): Bit.Option;
      OptionGroup(params?: Bit.OptionGroupParams): Bit.OptionGroup;
    };

    export const Blocks: {
      Actions(params?: Block.ActionsParams): Block.Actions;
      Context(params?: Block.ContextParams): Block.Context;
      Divider(params?: Block.DividerParams): Block.Divider;
      File(params?: Block.FileParams): Block.File;
      Header(params?: Block.HeaderParams): Block.Header;
      Image(params?: Block.ImageParams): Block.Image;
      Input(params?: Block.InputParams): Block.Input;
      Section(params?: Block.SectionParams): Block.Section;
    };

    export const Elements: {
      Button(params?: Element.ButtonParams): Element.Button;
      ChannelMultiSelect(
        params?: Element.ChannelMultiSelectParams
      ): Element.ChannelMultiSelect;
      ChannelSelect(
        params?: Element.ChannelSelectParams
      ): Element.ChannelSelect;
      Checkboxes(params?: Element.CheckboxesParams): Element.Checkboxes;
      ConversationSelect(
        params?: Element.ConversationSelectParams
      ): Element.ConversationSelect;
      DatePicker(params?: Element.DatePickerParams): Element.DatePicker;
      ExternalMultiSelect(
        params?: Element.ExternalMultiSelectParams
      ): Element.ExternalMultiSelect;
      ExternalSelect(
        params?: Element.ExternalSelectParams
      ): Element.ExternalSelect;
      Img(params?: Element.ImgParams): Element.Img;
      OverflowMenu(params?: Element.OverflowMenuParams): Element.OverflowMenu;
      RadioButtons(params?: Element.RadioButtonsParams): Element.RadioButtons;
      StaticMultiSelect(
        params?: Element.StaticMultiSelectParams
      ): Element.StaticMultiSelect;
      StaticSelect(params?: Element.StaticSelectParams): Element.StaticSelect;
      TextInput(params?: Element.TextInputParams): Element.TextInput;
      UserMultiSelect(
        params?: Element.UserMultiSelectParams
      ): Element.UserMultiSelect;
      UserSelect(params?: Element.UserSelectParams): Element.UserSelect;
    };

    export function HomeTab(params?: Surface.HomeTabParams): Surface.HomeTab;
    export function Message(params?: Surface.MessageParams): Surface.Message;
    export function Modal(params?: Surface.ModalParams): Surface.Modal;
  }

  export = SlackBlockBuilder;
}
```

**Comment by geoffp** [comment]
Nice work @karoun! I'm curious what version of TS you're using. Mine is 4.0.2, and it's complaining about the `declare namespace`s. It claims you shouldn't write those if you're already inside a `declare module`.

**Comment by karoun** [comment]
Oops, I didn't notice that because I have `skipLibCheck: true` in my TSConfig. Turns out the `declare` token isn't necessary in the first place — thanks @geoffp! I've updated the definitions [in my previous comment](https://github.com/raycharius/slack-block-builder/issues/24#issuecomment-708178283) to reflect that.

**Comment by geoffp** [comment]
@karoun they're working well for me so far. My eslint is still griping about `Array<T>` typing being forbidden in favor of `T[]`, but it does compile and run.

**Comment by karoun** [comment]
Awesome! Yeah, my ESLint gripes about the opposite 😅

**Comment by geoffp** [comment]
Everyone has their _very strong opinions_. 😄

**Comment by geoffp** [comment]
@raycharius @karoun would you like me to work this up into a PR?

**Comment by karoun** [comment]
That'd be sweet! 🙏 One thing I'm realizing, preserving the docblocks is important so typeahead gives meaningful autocomplete in IDEs that support it. Without it, users need to have the [reference docs](https://www.blockbuilder.dev/#/?id=start) open all the time.

**Comment by raycharius** [comment]
This is great! I've seen the PR, will get to this tomorrow evening, along with adding the `TimePicker` object! 🙌🏽

**Comment by raycharius** [comment]
Added in 1.7.0

Thanks @karoun and @geoffp!

**Comment by raycharius** [comment]
Providing an update on type definitions – they've been improved, including the JSDoc comments and IDE hints mentioned above. Currently available in v2.0.0 Beta.

``` bash
npm install --save slack-block-builder@2.0.0-beta.1
```
